### DISCUSSION TITLE
Use the new `once` helper for class helpers like `class_uses_recursive`

### DISCUSSION THREAD

**Discussion by Tofandel** [laravel/framework#51565] [Ideas]
Those are very hungry methods which call multiple array methods and internal reflection methods

I think it's a good idea to wrap the result in the new `once` helper as they have a static output because classes cannot be redefined in php

While for some cases it will cache the result in memory that will only be used once, if it's used multiple times, it could improve performances quite a bit

Thoughts? Maybe a benchmark on memory/speed to see if worth it?

**Comment by henzeb** [comment]
You can wrap them in a once when you need it... What would be the problem with that?

**Reply by Tofandel** [reply]
That wouldn't work because once needs to be in the same function to cache correctly and would be very clunky as well as unreliable in another wrapper because then the cache is only for your wrapper and not useful, not for the core of laravel and other plugins which also use this method

**Reply by henzeb** [reply]
class_uses_recursive is not going to calculate every possible list. so it's not like that you say class_uses_recursive(MyClass::class) and that somewhere else this is called with something else that it can use that cached information. when you do class_uses_recursive(MyClass::class) it will get you a list of all traits used recursively of MyClass::class, if this is being called elsewhere with class_use_recursive(SomeOtherClass::class) it won't use the once cached result, because it's a different request.

And I doubt it's good code if you have to call class_uses_recursive(MyClass::class)  in two completely different parts of your application, that somehow is called after each other.

**Reply by Tofandel** [reply]
Its not about different requests, it's about it being used in different places. The whole class uses recursive is bad design in the first place, when you have interfaces that work so much better, but here we are stuck with this unperformant solution

Funny you should mention

> And I doubt it's good code if you have to call class_uses_recursive(MyClass::class) in two completely different parts of your application, that somehow is called after each other.

Because that's exactly what happens in the core in a few places, it's very often used on any models multiple times, so I don't see why this shouldn't be cached

**Reply by henzeb** [reply]
but MyModel is not the same as YourModel. class_uses_recursive(MyModel::class) will not use the cached values from class_uses_recursive(YourModel::class).

**Reply by Tofandel** [reply]
Thanks for pointing something so obvious. So simple to run a search https://github.com/search?q=repo%3Alaravel%2Fframework%20class_uses_recursive&type=code

You would see that it's used extensively in the framework and called multiple times for the same classes in very different code spaces

It's also in the boot method of every model and relation so on average called 3 times per model

Is it so uncommon to use instanceof in separate places in your code base for the same class? Because this is what this method is but for traits. If not then you have a design issue and your code is not generic enough

**Reply by henzeb** [reply]
Well, What I am trying to do with my questions, is to get YOU to be more expressive. Not everyone here knows he laravel source code by heart. you want something, you have to explain why.

instanceof is less heavy then recursively loop through all parent classes, if it was, I'd make a private "isInstanceOfMyClass() method, either using once, or my own caching implementation.
